File: 백준_3 Gold/16724_피리 부는 사나이.js
const fs = require("fs");
const inp = fs.readFileSync("/dev/stdin").toString().split("\n");
const [N, M] = inp[0].split(" ").map((v) => parseInt(v));
const board = inp.splice(1);

let ans = 0;
const visited = Array.from({ length: N }, () => Array.from({ length: M }, () => 0));

const dirMap = {
  U: [-1, 0],
  D: [1, 0],
  L: [0, -1],
  R: [0, 1],
};

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    if (visited[i][j] == 2) continue;

    const stack = [[i, j]];
    while (true) {
      const cur = stack[stack.length - 1];
      if (visited[cur[0]][cur[1]] == 0) {
        visited[cur[0]][cur[1]] = 1;
        stack.push([
          cur[0] + dirMap[board[cur[0]][cur[1]]][0],
          cur[1] + dirMap[board[cur[0]][cur[1]]][1],
        ]);
        continue;
      }
      if (visited[cur[0]][cur[1]] == 1) ans++;
      for (let k = 0; k < stack.length; k++) visited[stack[k][0]][stack[k][1]] = 2;
      break;
    }
  }
}

console.log(ans);

File: 백준_3 Gold/16724_피리 부는 사나이.py

N, M = map(int, input().split())
board = [input() for _ in range(N)]

visited = [[0 for j in range(M)] for i in range(N)]

dirMap = {
    'U' : [-1,  0],
    'D' : [ 1,  0],
    'L' : [ 0, -1],
    'R' : [ 0,  1],
}

ans = 0

def findCycleOrEnd(i, j, path):
    if 0 > i or i >= N or 0 > j or j >= M: return [[i, j]]

    if visited[i][j] == 2: return []
    if visited[i][j] == 1:
        cycle = [[i, j]]
        while path[-1][0] != i or path[-1][1] != j:
            cycle.append(path.pop())
        return cycle

    visited[i][j] = 1
    path.append([i, j])
    curD = dirMap[board[i][j]]
    cycle = findCycleOrEnd(i + curD[0], j + curD[1], path)
    visited[i][j] = 2

    return cycle

ans = 0
for i in range(N):
    for j in range(M):
        if visited[i][j] == 2: continue
        curCycle = findCycleOrEnd(i, j, [])

        if len(curCycle) > 0:
            ans += 1

print(ans)